File: elementary/71-simplify-path.java
package elementary;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

class Solution {
    public String simplifyPath(String path) {
        Stack<String> stack = new Stack<>();
        String[] ss = path.split("/");
        
        for (String s : ss) {
            if (!stack.empty() && s.equals("..")) {
                stack.pop();
            } else if (!s.equals(".") && !s.equals("") && !s.equals("..")) {
                stack.push(s);
            }
        }
        
        List<String> list = new ArrayList<>(stack);
        return "/" + String.join("/", list);
    }
}

File: elementary/71-simplify-path.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    string simplifyPath(string path) {
        vector<string> stack;
        int n = path.length();

        for (int i = 0; i < n;) {
            while (i < n && path[i] == '/') i++;
            if (i == n) break;
            int start = i;

            while (i < n && path[i] != '/') i++;
            int end = i - 1;

            string s = path.substr(start, end - start + 1);

            if (s == "..") {
                if (!stack.empty()) stack.pop_back();
            } else if (s != ".") {
                stack.push_back(s);
            }
        }

        if (stack.empty()) return "/";
        
        string res = "";
        for (auto s : stack) {
            res += "/" + s;
        }
        return res;
    }
};